function-controller/pkg/controller/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright 2018-Present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package autoscaler
+
+import (
+	"github.com/projectriff/riff/message-transport/pkg/transport/metrics"
+	"fmt"
+	"sync"
+	"time"
+	"math"
+	"log"
+	"io"
+	"github.com/projectriff/riff/message-transport/pkg/transport"
+)
+
+//go:generate mockery -name=AutoScaler -output mockautoscaler -outpkg mockautoscaler
+
+type AutoScaler interface {
+	// Set maximum replica count policy.
+	SetMaxReplicasPolicy(func(topic string, function string) int)
+
+	// Set delay scale down policy.
+	SetDelayScaleDownPolicy(func(function string) time.Duration)
+
+	// Run starts the autoscaler receiving and sampling metrics.
+	Run()
+
+	// Close stops the autoscaler receiving and sampling metrics.
+	io.Closer
+
+	// InformFunctionReplicas is used to tell the autoscaler the actual number of replicas there are for a given
+	// function. The function is not necessarily being monitored by the autoscaler.
+	InformFunctionReplicas(function FunctionId, replicas int)
+
+	// StartMonitoring starts monitoring metrics for the given topic and function.
+	StartMonitoring(topic string, function FunctionId) error
+
+	// StopMonitoring stops monitoring metrics for the given topic and function.
+	StopMonitoring(topic string, function FunctionId) error
+
+	// Propose proposes the number of replicas for functions that are being monitored.
+	Propose() map[FunctionId]int
+}
+
+// FunctionId identifies a function in the default namespace.
```

Comment:
Not sure "in the default namespace" is accurate. It's more that it doesn't take namespace into account (which should be ultimately remedied)

Reply:
Good point. I've removed the namespace remark and added a TODO.

For the record: I haven't yet gotten my head around what's needed to support namespaces.